apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex

```diff
@@ -0,0 +1,94 @@
+  require Logger
+
+  def get_exchange_rate_behaviour, do: Application.get_env(:walletapi, :exchange_rate_behaviour)
+
+  defp get_quotes({:error, :not_found}, _timestamp), do: {:error, :not_found}
+
+  defp get_quotes(response, pair) do
+    case response["quotes"][pair] do
+      nil -> {:error, :not_found}
+      _ -> Decimal.from_float(response["quotes"][pair])
+    end
+  end
+
+  def get_exchange_rate(source_currency_code, currency_code, timestamp) do
+    if(!currency_code) do
+      raise DisplayError, message: "No currency code specified"
+    end
+
+    pair = source_currency_code <> currency_code
+    response = get_exchange_rate_behaviour().query_exchange_rate(source_currency_code, currency_code, timestamp)
+    get_quotes(response, pair)
+  end
+
+  def query_exchange_rate(source_currency_code, _currency_code, timestamp) do
+    time_in_ms =
+      if timestamp === nil do
+        DateTime.utc_now()
+      else
+        DateTime.from_unix!(timestamp, :millisecond)
+      end
+
+    # is08601 format : YYYY-MM-DD
+    date = Date.to_iso8601(time_in_ms)
+
+    # if given time within last 24 hours - get current exchange rates from API.
+    if DateTime.diff(DateTime.utc_now(), time_in_ms) < 24 * 60 * 60 do
+      get_data_from_api(source_currency_code, date)
+    else
+      get_data_from_cache(source_currency_code, date)
+    end
+  end
+
+  defp get_data_from_cache(source_currency_code, date) do
+    # global_ttl set to 1 year to store historical exchange rates
+    # *****************TODO******************
+    # Concache is in-memory cache and gets reset if the pod is cleaned
+    # Add another cache layer (GoogleCloud Cache) to store the values
+    # If in-memory cache is cleaned, get the rates from GoogleClousCache
+    # Sequence : ConCache (cache miss)-> GoogleCloudCache (cache miss)-> CurrencyLayerAPI
+    # ***************************************
+    data_from_cache = ConCache.get(:exchange_rate_cache, source_currency_code <> "-" <> date)
+
+    # cache miss get the data from the api
+    if data_from_cache == nil do
+      get_data_from_api(source_currency_code, date)
+    else
+      data_from_cache
+    end
+  end
+
+  defp return_response(source_currency_code, date, response) do
+    if Map.get(response, "success") == false do
+      IO.puts("Failed to fetch Exchange Rates")
+      Logger.error(fn -> "Failed to fetch Exchange Rates: #{inspect(response)}" end)
+      {:error, :not_found}
+    else
+      ConCache.put(:exchange_rate_cache, source_currency_code <> "-" <> date, response)
+      response
+    end
+  end
+
+  defp get_data_from_api(source_currency_code, date) do
+    path = "/historical?"
+    access_key = System.get_env("CURRENCY_LAYER_ACCESS_KEY")
+    req_body = URI.encode_query(access_key: access_key, source: source_currency_code, date: date)
+
+    case HTTPoison.get(@api_url <> path <> req_body) do
+      {:ok, %HTTPoison.Response{body: body, status_code: 200}} ->
+        response = Poison.decode!(body)
```

Comment:
Just curious how it works in Elixir world, if you call a bang-function, are you supposed to handle the raised exception? And if you don't handle it here, are you supposed to add an exclamation mark to your own function or not?

Reply:
Even if we use bang_function we still need to catch the exception. And the problem with the currency_layer api is that it returns a 200 status even for incorrect request hence additional checks are performed in the return_response() function to validate the response.